velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,327 @@
+/// (not a "value_type" but a memory buffer store normalized keys) for memcmp
+/// and memcpy operators, means not ValueSwappable.
+/// This is also the reason we can not use std::sort directly, and implement
+/// quickSort in PrefixSortContext.
+struct PrefixSortIterator {
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix_(prefix), entrySize_(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
+      : prefix_(other.prefix_), entrySize_(other.entrySize_) {}
+
+  // Inline functions should be place in include file to avoid link error.
+  FOLLY_ALWAYS_INLINE char* operator*() const {
+    return prefix_;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator++() {
+    prefix_ += entrySize_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator--() {
+    prefix_ -= entrySize_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator++(int) {
+    auto tmp = *this;
+    prefix_ += entrySize_;
+    return tmp;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator--(int) {
+    auto tmp = *this;
+    prefix_ -= entrySize_;
+    return tmp;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator+(const uint64_t i) const {
+    auto result = *this;
+    result.prefix_ += i * entrySize_;
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator-(const uint64_t i) const {
+    PrefixSortIterator result = *this;
+    result.prefix_ -= i * entrySize_;
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator=(
+      const PrefixSortIterator& other) {
+    prefix_ = other.prefix_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE uint64_t
+  operator-(const PrefixSortIterator& other) const {
+    return (prefix_ - other.prefix_) / other.entrySize_;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator<(const PrefixSortIterator& other) const {
+    return prefix_ < other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator>(const PrefixSortIterator& other) const {
+    return prefix_ > other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator>=(const PrefixSortIterator& other) const {
+    return prefix_ >= other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator<=(const PrefixSortIterator& other) const {
+    return prefix_ <= other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator==(const PrefixSortIterator& other) const {
+    return prefix_ == other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator!=(const PrefixSortIterator& other) const {
+    return prefix_ != other.prefix_;
+  }
+
+ private:
+  char* prefix_;
+  const uint64_t entrySize_;
+};
+
+/// Provides a series of methods (mostly required by the sort
+/// algorithm) for PrefixSort. Maintains the buffer for swap operations and
+/// some necessary context information such as entrySize.
+class PrefixSortContext {
+ public:
+  PrefixSortContext(uint64_t entrySize, char* swapBuffer);
```

Comment:
Let's document this constructor to clarify that swapBuffer must be at least entrySize bytes long.

Reply:
@skadilover I see this 'resolved', but I don't see comments in the code. Am I missing something?

Reply:
No 
I missing something , I added the doc at class...

Fixed